### PR TITLE
[Warnings]Make it possible to add reason with unwarn 

### LINKED ISSUE #3093: `[p]unwarn` has no arg for reason, and modlog says one is addable
# Command bugs

#### Command name

`[p]unwarn`

#### What cog is this command from?

`warnings`

#### What were you expecting to happen?

There to be a third argument where you can put an optional reason for unwarning the user.

#### What actually happened?

There is no argument that allows you to set the reason. The modlog channel says "**Reason:** Use the `reason` command to add it", implying that you should be able to set a reason without needing to do it in two commands.

#### How can we reproduce this issue?

1. `[p]load warnings`
2. Set up a modlog channel, `[p]modlogset modlog <channel>`
3. `[p]warn` someone
4. `[p]unwarn` them, and note how you cannot add a reason
5. Notice how the modlog says that there is no reason added to the case
6. Use `[p]reason` and notice how you can add a warning to the case

#### Others
I may (attempt!) to PR a fix myself - though I'm a little busy so feel free to do this yourself.
I would like to thank [Skyrider#1337 for reporting this bug](https://discordapp.com/channels/133049272517001216/387398816317440000/639794017596473355). And yes, I waited a whole 11.5 hours for them to open an issue themselves.

### PR DESCRIPTION
**Pull request by Dav-Git**
Type

Bugfix
Enhancement

[X] New feature

Description of the changes

This closes #3093 .
This overwrites #3441 because I'm stupid and deleted the fork.

### CODE REVIEW

File: redbot/cogs/warnings/warnings.py
Comment (Drapersniper) [inline redbot/cogs/warnings/warnings.py]: Not sure if I'm missing anything here but is there a purpose to be running this string through the translator?

File: redbot/cogs/warnings/warnings.py
Comment (Drapersniper) [inline redbot/cogs/warnings/warnings.py]:
```suggestion
                        "Run `{prefix}warningset allowcustomreasons true` to enable custom "
```

File: redbot/cogs/warnings/warnings.py
Comment (Drapersniper) [inline redbot/cogs/warnings/warnings.py]:
```suggestion
                msg = _("That is not a registered reason.")
```
Please no shouting is too early in the morning.

File: redbot/cogs/warnings/warnings.py
Comment (Drapersniper) [inline redbot/cogs/warnings/warnings.py]: What is the purpose of line 462 to 464?

**Review by Drapersniper — DISMISSED** [review]

File: redbot/cogs/warnings/warnings.py
Comment (Dav-Git) [reply redbot/cogs/warnings/warnings.py]: You're 100% correct, there isn't. This one slipped through.

**Review by Drapersniper — COMMENTED** [review]
This is a review for the Code and not functionality:
LGTM (Can't check functionality atm)

File: redbot/cogs/warnings/warnings.py
Comment (Drapersniper) [inline redbot/cogs/warnings/warnings.py]: Actually stating the valid reasons, when `custom_allowed` is disallowed and `reason.lower() not in registered_reasons`  would be more user friendly (something along the lines of "Please use one of the following: ...")

Right now `if reason.lower() not in registered_reasons and not custom_allowed` this just fails silently

Comment (Dav-Git) [reply redbot/cogs/warnings/warnings.py]: Will look at this when I return home later today

Comment (Dav-Git) [reply redbot/cogs/warnings/warnings.py]: This would then require to be edited in the warn command as well. My knowledge in string formatting is limited (everyone tells me my code is to slow) that I'd rather leave that up to someone else as it also doesn't really fit in this PR in my opinion. If you really want to see this included I'll need some guidance via discord.

I'm guessing that ``reason_lower`` is a list of strings that in one way ore another can be concatinated and passed to an embed which is then sent to the user. If we exceed the embed limit with the amount of registered reasons that's going to require some logic I don't know how to code and I'm not really sure if it's worth it at that point. Leaving it with just the instructions on how to enable custom warnings seems reasonable to me.

**Comment by Dav-Git** [comment]
Are they supposed to not report anything? Do I need to do something?

File: redbot/cogs/warnings/warnings.py
Comment (Drapersniper) [inline redbot/cogs/warnings/warnings.py]: I still think we should return a user friendly message when a valid reason is not present, just silently doing nothing is very confusing imo

Comment (Drapersniper) [reply redbot/cogs/warnings/warnings.py]: If another QA/Core-dev disagrees let me know and i'll review this as it is.

**Comment by Jackenmen** [comment]
I still think that registered warn reasons shouldn't be used for unwarn - they will simply not make sense as unwarn reasons. *I think* allowing any text as unwarn reason is enough but I guess there's also the option of registering unwarn reasons.

File: redbot/cogs/warnings/warnings.py
Comment (Jackenmen) [reply redbot/cogs/warnings/warnings.py]: I do agree we shouldn't fail silently, though I don't think we should list all registered reasons, especially not in scope of this PR.

**Comment by Dav-Git** [comment]
I think just removing the use of registered reasons for unwarn is the way forward with this. You guys let me know if that's what you want me to do.

**Review by Drapersniper — COMMENTED** [review]
Okay now that we return a message when reason_type is None I’m more okay with this. Although jack made a point And not sure how strongly he feels about this one @jack1142 do you think removing the usage of the built in reason type is important enough here?

**Comment by Jackenmen** [comment]
@Dav-Git after additional discussion, we decided that `[p]uwarn` should just allow any reason.

File: redbot/cogs/warnings/warnings.py
Comment (Jackenmen) [inline redbot/cogs/warnings/warnings.py]: This can just be (though this part might not even exist after allowing any reason):
```suggestion
            reason_msg = reason_type["description"]
```

**Comment by Dav-Git** [comment]
Sounds good to me. I'll remove it for both commands. Maybe the commands to set the warn reasons are pending removal too though? Especially since the reasons then no longer server a purpose.

**Comment by Dav-Git** [comment]
Correction, I misread that, allowing any reason for just unwarn

File: redbot/cogs/warnings/warnings.py
Comment (Jackenmen) [inline redbot/cogs/warnings/warnings.py]: This is unneeded, you can just use the `reason` variable instead of `reason_msg` in `modlog.create_case` call.

File: redbot/cogs/warnings/warnings.py
Comment (Jackenmen) [inline redbot/cogs/warnings/warnings.py]: Reason should be optional:
```suggestion
        self,
        ctx: commands.Context,
        user: Union[discord.Member, int],
        warn_id: str,
        reason: str = None,
```

**Review by Jackenmen — CHANGES_REQUESTED** [review]
Looks like I'm gonna review this PR after all, only one change should be needed here before approval :)

File: redbot/cogs/warnings/warnings.py
Comment (Dav-Git) [reply redbot/cogs/warnings/warnings.py]: Where is the change?

Comment (Dav-Git) [reply redbot/cogs/warnings/warnings.py]: oh i'm blind. yeah, agreed

**Review by Jackenmen — APPROVED** [review]
LGTM

**Pull request merged**
